### PR TITLE
feat(VncConsole): More possibilities to customize component

### PR DESCRIPTION
**Pull request by bond95**
**What**:

Added possibility to extract toolbar to any other component. Added scaling prop to VncConsole, and
enhance `textConnecting` props to use node type as well. And fixed disconnecting.

We need this features for console page in oVirt VM Portal: https://github.com/oVirt/ovirt-web-ui/pull/956

Example from oVirt VM Portal:
image

@mareklibra please, review

### CODE REVIEW

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js [hunk 1]
```diff
@@ -62,17 +63,17 @@ class VncConsole extends React.Component {
   }
 
   componentWillUnmount() {
-    this.removeEventListeners();
     this.disconnect();
+    this.removeEventListeners();
+    this.rfb = undefined;
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: With this change, the `RFB.disconnect()` is skipped. Can you please explain reason for that?

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js [hunk 2]
```diff
@@ -62,17 +63,17 @@ class VncConsole extends React.Component {
...
   }
 
-  disconnect() {
+  disconnect = () => {
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: Why this change, please?

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js [hunk 3]
```diff
@@ -1,30 +1,60 @@
...
-    </MenuItem>
-  </Dropdown>
-);
+const VncActions = ({
+  textSendShortcut,
+  textCtrlAltDel,
+  textDisconnect,
+  onCtrlAltDel,
+  onDisconnect,
+  portalToolbarTo
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: The name is too specific for the `oVirt VM Portal`

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js [hunk 4]
```diff
@@ -1,30 +1,60 @@
...
+        {textDisconnect}
+      </Button>
+    </div>
+  );
+
+  if (!portalToolbarTo) {
+    return toolbar;
+  }
+  return (
+    document.getElementById(portalToolbarTo) && ReactDOM.createPortal(toolbar, document.getElementById(portalToolbarTo))
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: I don't think we should perform such manipulations. I believe change in the component composition should be preferred over this approach.

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js [hunk 5]
```diff
@@ -1,30 +1,60 @@
...
-  textSendShortcut: PropTypes.string
+  textSendShortcut: PropTypes.string,
+  textDisconnect: PropTypes.string,
+
+  portalToolbarTo: PropTypes.string
 };
 
 VncActions.defaultProps = {
   onCtrlAltDel: noop,
+  onDisconnect: noop,
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: This is a breaking change.
The consuming application is **silently** required to add implementation. Otherwise the button just appears there without any function.

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js [hunk 6]
```diff
@@ -95,11 +96,11 @@ class VncConsole extends React.Component {
     this.props.onSecurityFailure(e);
   };
 
-  removeEventListeners() {
+  removeEventListeners = () => {
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: likewise

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js [hunk 7]
```diff
@@ -143,11 +154,21 @@ class VncConsole extends React.Component {
     return (
       <div className={classNames('vnc-console', this.props.topClassName)}>
         {this.props.children}
-        <Toolbar.RightContent>{rightContent}</Toolbar.RightContent>
-        <Toolbar.Results>
-          {status}
-          {this.novncStaticComponent}
-        </Toolbar.Results>
+        {portalToolbarTo ? (
```
Comment (mareklibra) [inline packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: as commented above

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js (hunk 1 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: But no, it doesn't skipped. Line 66 perform disconnecting. And previously `onDisconnected` event doesn't work, because it removed all listeners before it disconnected.

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js (hunk 3 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: Well, it calling `portalToolbarTo`, because if it set it do portal toolbar to some element, not because this is from VM Portal. But yeah, it better to rename.

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js (hunk 4 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: Yes, but in that case appear problem with customization. For example, in VM Portal we need only one toolbar not two and we can't `Send Key` dropdown move to our toolbar, because console is fully under control of `VncConsole` component.

And I think this is most suitable solution.

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js (hunk 5 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: Ok, I'll do `onDisconnect` props required, but this is was just copy from `onCtlAltDel`. So why in this case `onCtrlAltDel: noop` not required too?

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js (hunk 6 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: Because of context.

**Comment by bond95** [comment]
@mareklibra please, review.

**Comment by jeff-phillips-18** [comment]
@mareklibra Could you please take another look?

**Comment by bond95** [comment]
@priley86 could you, please, look at this PR?

File: packages/patternfly-3/react-console/src/VncConsole/VncConsole.js (hunk 2 above)
Comment (priley86) [reply packages/patternfly-3/react-console/src/VncConsole/VncConsole.js]: +1

**Comment by priley86** [comment]
like other VncConsole PRs, this is hard to visualize. Can you possibly share an environment this is running w/ me? Or maybe it is easier to record a gif from your machine..

Also, do you plan to add PF4 variation? This is still using PF3 menus/buttons/etc..

**Comment by bond95** [comment]
> like other VncConsole PRs, this is hard to visualize. Can you possibly share an environment this is running w/ me? Or maybe it is easier to record a gif from your machine..

captured (4)

> Also, do you plan to add PF4 variation? This is still using PF3 menus/buttons/etc..

Yes, but in separate PR.

**Comment by bond95** [comment]
@priley86 Could you please take another look?

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js [hunk 8]
```diff
@@ -1,30 +1,59 @@
...
+        {textDisconnect}
+      </Button>
+    </div>
+  );
+
+  if (!insertToolbarTo) {
+    return toolbar;
+  }
+  return (
+    document.getElementById(insertToolbarTo) && ReactDOM.createPortal(toolbar, document.getElementById(insertToolbarTo))
```
Comment (priley86) [inline packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: does this need to be a portal? We don't typically use these for elements that will render in the typical DOM/page content area (they are more often used for appending modals/popovers/etc.). 

I'm not currently using this downstream, so I can't say for sure, but you may be able to just expose it as a simple stateless component...

File: packages/patternfly-3/react-console/src/VncConsole/__snapshots__/VncActions.test.js.snap [hunk 9]
```diff
@@ -1,43 +1,65 @@
...
 exports[`VncActions renders correctly component hierarchy 1`] = `
-<DropdownButton
-  bsStyle="default"
-  id="console-send-shortcut"
-  onClick={[Function]}
-  title="My Send Shortcut description"
->
-  <MenuItem
-    bsClass="dropdown"
+<div>
```
Comment (priley86) [inline packages/patternfly-3/react-console/src/VncConsole/__snapshots__/VncActions.test.js.snap]: additional div?

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js [hunk 10]
```diff
@@ -1,30 +1,59 @@
...
 VncActions.propTypes = {
+  onDisconnect: PropTypes.func.isRequired,
   onCtrlAltDel: PropTypes.func,
 
   textCtrlAltDel: PropTypes.string,
-  textSendShortcut: PropTypes.string
+  textSendShortcut: PropTypes.string,
+  textDisconnect: PropTypes.string,
+
+  insertToolbarTo: PropTypes.string
```
Comment (jeff-phillips-18) [inline packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: When would this be used? I don't see it being used here at all. 
These propTypes should all be commented.

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js (hunk 8 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: > does this need to be a portal?

Yes, this component is depend on VncConsole component and for right work should only be used inside of VncConsole. So from my point of view portals are the best solution.

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js (hunk 10 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: > When would this be used? I don't see it being used here at all.

It is used in VncActions.

https://github.com/patternfly/patternfly-react/blob/223e582264d215dd6341bd87507a2774b4b0e4b0/packages/patternfly-3/react-console/src/VncConsole/VncActions.js#L30-L35

File: packages/patternfly-3/react-console/src/VncConsole/__snapshots__/VncActions.test.js.snap (hunk 9 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/__snapshots__/VncActions.test.js.snap]: No, `div` should be there. https://github.com/patternfly/patternfly-react/blob/223e582264d215dd6341bd87507a2774b4b0e4b0/packages/patternfly-3/react-console/src/VncConsole/VncActions.js#L17-L28

File: packages/patternfly-3/react-console/src/VncConsole/VncActions.js (hunk 10 above)
Comment (bond95) [reply packages/patternfly-3/react-console/src/VncConsole/VncActions.js]: > These propTypes should all be commented.

Added comment.

**Comment by bond95** [comment]
@priley86 could you, please, look one more?

**Comment by priley86** [comment]
merging - please be sure to file any follow up issues after testing downstream further.

thanks @bond95 !
